File: apps/web/lib/documents.ts
/**
 * API client for documents: list, get, patch, create from template, and search.
 * Document responses use snake_case; search responses use camelCase.
 */

const getBase = () =>
  process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

// ---- Response types (match backend) ----

export type DocumentSummary = {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  version: number;
  party_id: string | null;
  party_name: string | null;
  doc_type: string | null;
};

export type Document = DocumentSummary & {
  content: string;
};

// ---- PATCH request (range-based replace) ----

export type ReplaceRange = {
  start: number;
  end: number;
};

export type DocumentChange = {
  operation: "replace";
  range: ReplaceRange;
  text: string;
};

// ---- Search response (backend returns camelCase) ----

export type SearchMatch = {
  documentId: string;
  title: string;
  snippet: string;
  startIndex: number;
  endIndex: number;
};

export type SearchResponse = {
  matches: SearchMatch[];
  total: number;
};

// ---- API errors (shared error shape from backend) ----

export type ApiError = {
  error: string;
  code: number;
};

async function parseError(res: Response): Promise<ApiError> {
  let body: ApiError;
  try {
    body = await res.json();
  } catch {
    body = { error: res.statusText || "Request failed", code: res.status };
  }
  return body;
}

// ---- listDocuments ----

export type ListDocumentsFilters = {
  party_id?: string;
  party_name?: string;
  doc_type?: string;
};

/**
 * List document summaries (no content). Optional filters: party_id, party_name, doc_type.
 */
export async function listDocuments(
  filters: ListDocumentsFilters = {}
): Promise<DocumentSummary[]> {
  const params = new URLSearchParams();
  if (filters.party_id != null) params.set("party_id", filters.party_id);
  if (filters.party_name != null) params.set("party_name", filters.party_name);
  if (filters.doc_type != null) params.set("doc_type", filters.doc_type);
  const qs = params.toString();
  const url = `${getBase()}/documents${qs ? `?${qs}` : ""}`;
  const res = await fetch(url);
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<DocumentSummary[]>;
}

// ---- createFromTemplate ----

/**
 * Create a new document as a copy of a template. Returns the new document (use its id for editing).
 * Throws on 404 if template not found.
 */
export async function createFromTemplate(
  templateId: string,
  title?: string
): Promise<Document> {
  const url = `${getBase()}/documents/from-template`;
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ templateId, title }),
  });
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<Document>;
}

// ---- getDocument ----

/**
 * Fetch full document by id. Returns null if 304 (Not Modified).
 * Throws on 4xx/5xx with ApiError-shaped message.
 */
export async function getDocument(
  id: string,
  ifNoneMatchVersion?: number
): Promise<Document | null> {
  const url = `${getBase()}/documents/${encodeURIComponent(id)}`;
  const headers: HeadersInit = {};
  if (ifNoneMatchVersion != null) {
    headers["If-None-Match"] = String(ifNoneMatchVersion);
  }
  const res = await fetch(url, { headers });
  if (res.status === 304) return null;
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<Document>;
}

// ---- patchDocument ----

/**
 * Apply range-based changes and/or title update. Sends If-Match: version.
 * Returns updated document. Throws on 404, 412 (version mismatch), 400 (e.g. out of range).
 */
export async function patchDocument(
  id: string,
  version: number,
  changes: DocumentChange[],
  options?: { title?: string }
): Promise<Document> {
  const url = `${getBase()}/documents/${encodeURIComponent(id)}`;
  const body: { changes: DocumentChange[]; title?: string } = { changes };
  if (options?.title !== undefined) body.title = options.title;
  const res = await fetch(url, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
      "If-Match": String(version),
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<Document>;
}

// ---- deleteDocument ----

/**
 * Delete document by id. Throws on 404.
 */
export async function deleteDocument(id: string): Promise<void> {
  const url = `${getBase()}/documents/${encodeURIComponent(id)}`;
  const res = await fetch(url, { method: "DELETE" });
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
}

// ---- searchDocuments ----

export type SearchDocumentsOptions = {
  limit?: number;
  offset?: number;
  documentIds?: string[];
};

/**
 * Search all documents (or a subset by documentIds). Returns matches and total count.
 * q is required; limit/offset for pagination.
 */
export async function searchDocuments(
  q: string,
  options: SearchDocumentsOptions = {}
): Promise<SearchResponse> {
  const params = new URLSearchParams();
  params.set("q", q.trim());
  if (options.limit != null) params.set("limit", String(options.limit));
  if (options.offset != null) params.set("offset", String(options.offset));
  if (options.documentIds?.length) {
    options.documentIds.forEach((id) => params.append("document_ids", id));
  }
  const url = `${getBase()}/documents/search?${params.toString()}`;
  const res = await fetch(url);
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<SearchResponse>;
}

// ---- searchInDocument ----

export type SearchInDocumentOptions = {
  limit?: number;
  offset?: number;
};

/**
 * Search within a single document. Same response shape as searchDocuments.
 */
export async function searchInDocument(
  docId: string,
  q: string,
  options: SearchInDocumentOptions = {}
): Promise<SearchResponse> {
  const params = new URLSearchParams();
  params.set("q", q.trim());
  if (options.limit != null) params.set("limit", String(options.limit));
  if (options.offset != null) params.set("offset", String(options.offset));
  const url = `${getBase()}/documents/${encodeURIComponent(docId)}/search?${params.toString()}`;
  const res = await fetch(url);
  if (!res.ok) {
    const err = await parseError(res);
    throw new Error(err.error || `HTTP ${err.code}`);
  }
  return res.json() as Promise<SearchResponse>;
}

File: apps/backend/app/routes/documents.py
"""Document routes: list, get by id, PATCH content, search. Thin HTTP layer; logic in service."""
from fastapi import APIRouter, Depends, Header, HTTPException, Path, Query
from fastapi.responses import JSONResponse, Response

from app.core.deps import get_document_service, get_search_service
from app.models.api import (
    CreateFromTemplateRequest,
    DocumentResponse,
    DocumentSummaryResponse,
    PatchDocumentRequest,
    SearchMatchResponse,
    SearchResponse,
)
from app.routes.utils.formatting import parse_version_header
from app.routes.utils.mappers import doc_to_response, summary_to_response
from app.services.documents import DocumentNotFoundError, DocumentService, PreconditionFailedError
from app.services.search import SearchService

router = APIRouter(prefix="/documents", tags=["documents"])

VERSION_HEADER = "Version"

## list documents (summaries only, no content). Optional query params filter by metadata.
@router.get("", response_model=list[DocumentSummaryResponse])
def list_documents(
    party_id: str | None = Query(None, description="Filter by party_id"),
    party_name: str | None = Query(None, description="Filter by party_name"),
    doc_type: str | None = Query(None, description="Filter by doc_type (e.g. template, loan_agreement)"),
    _service: DocumentService = Depends(get_document_service),
):
    """List documents (summaries only). Omit query params for all; use party_id, party_name, or doc_type to filter."""
    summaries = _service.list_document_summaries(
        party_id=party_id,
        party_name=party_name,
        doc_type=doc_type,
    )
    return [summary_to_response(s) for s in summaries]


## search (must be before /{doc_id} so /documents/search is not matched as doc_id)
@router.get("/search", response_model=SearchResponse)
def search_documents(
    q: str = Query(..., min_length=1, description="Search query (phrase, case-insensitive)"),
    limit: int = Query(10, ge=1, le=100),
    offset: int = Query(0, ge=0),
    document_ids: list[str] | None = Query(None, description="Restrict to these document IDs"),
    _service: SearchService = Depends(get_search_service),
) -> SearchResponse:
    """Search all documents (or a subset by document_ids). Returns matches and total count."""
    page, total = _service.search(q=q, limit=limit, offset=offset, document_ids=document_ids)
    matches = [
        SearchMatchResponse(
            document_id=m.document_id,
            title=m.title,
            snippet=m.snippet,
            start_index=m.start_index,
            end_index=m.end_index,
        )
        for m in page
    ]
    return SearchResponse(matches=matches, total=total)


@router.get("/{doc_id}/search", response_model=SearchResponse)
def search_in_document(
    doc_id: str = Path(..., description="Document ID"),
    q: str = Query(..., min_length=1, description="Search query"),
    limit: int = Query(10, ge=1, le=100),
    offset: int = Query(0, ge=0),
    _service: SearchService = Depends(get_search_service),
) -> SearchResponse:
    """Search within a single document. Same response shape as global search."""
    page, total = _service.search(q=q, limit=limit, offset=offset, document_ids=[doc_id])
    matches = [
        SearchMatchResponse(
            document_id=m.document_id,
            title=m.title,
            snippet=m.snippet,
            start_index=m.start_index,
            end_index=m.end_index,
        )
        for m in page
    ]
    return SearchResponse(matches=matches, total=total)


## create a new document from a template (copy); must be before /{doc_id}
@router.post("/from-template", status_code=201, response_model=DocumentResponse)
def create_from_template(
    body: CreateFromTemplateRequest,
    _service: DocumentService = Depends(get_document_service),
) -> JSONResponse:
    """Create a new document as a copy of the given template. Returns the new document (use its id for PATCH)."""
    try:
        doc = _service.create_document_from_template(
            body.template_id,
            title=body.title,
        )
    except DocumentNotFoundError:
        raise HTTPException(status_code=404, detail="Template not found")
    response_body = doc_to_response(doc).model_dump(mode="json")
    return JSONResponse(
        content=response_body,
        status_code=201,
        headers={VERSION_HEADER: str(doc.version)},
    )


## get full document by id
@router.get("/{doc_id}")
def get_document(
    doc_id: str = Path(..., description="Document ID"),
    if_none_match: str | None = Header(None, alias="If-None-Match"),
    _service: DocumentService = Depends(get_document_service),
) -> Response:
    """Get full document by id. 304 if If-None-Match matches current version."""
    doc = _service.get_document(doc_id)
    if doc is None:
        raise HTTPException(status_code=404, detail="Document not found")
    client_version = parse_version_header(if_none_match)
    if client_version is not None and client_version == doc.version:
        return Response(
            status_code=304,
            headers={VERSION_HEADER: str(doc.version)},
        )
    body = doc_to_response(doc).model_dump(mode="json")
    return JSONResponse(
        content=body,
        headers={VERSION_HEADER: str(doc.version)},
    )

## apply changes to document content
## will return 412 if version does not match, client will need to fetch the latest version and apply changes again
@router.patch("/{doc_id}")
def patch_document(
    doc_id: str = Path(..., description="Document ID"),
    body: PatchDocumentRequest = ...,
    if_match: str | None = Header(None, alias="If-Match"),
    _service: DocumentService = Depends(get_document_service),
) -> JSONResponse:
    """Apply changes to document content. Requires If-Match: version. 412 if missing or mismatch."""
    try:
        version = parse_version_header(if_match)
        doc = _service.update_document_content(
            doc_id, version, body.changes, title=body.title
        )
    except DocumentNotFoundError:
        raise HTTPException(status_code=404, detail="Document not found")
    except PreconditionFailedError:
        raise HTTPException(
            status_code=412,
            detail="Precondition Failed: If-Match is required and must match current document version.",
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    response_body = doc_to_response(doc).model_dump(mode="json")
    return JSONResponse(
        content=response_body,
        headers={VERSION_HEADER: str(doc.version)},
    )
